### DISCUSSION TITLE
Directus "server-side" functions

### DISCUSSION THREAD

**Discussion by erondpowell** [directus/directus#13261] [Feature Requests]
**Are we ever going to want an extension for Directus-side logic and validation?**

Like say I wanted to build a quiz. I could:
- grade all quizzes in the frontend- which poses problems.
- use a webhook to send the grades off to another server somewhere.
- use Data Flows? Since there are validation operations, you might be able to create an "answer key" in the database then check all incoming quiz results, then write in a good score for each answer that passed validation, and bad scores for answers that don't. 

For more advanced stuff, like using NLP to check student's writing performance, you'd probably want to use flask or an NLP cloud service. So a webhook is ideal there. But I'm wondering if there is a middle-ground where data flows reaches a limit, but still a little overkill to spin up and maintain a separate server.

Will we ever add a Module (or operation in data flows) to let us create "Directus-side" functions for transformation and validation directly? Also, would we actually want/need this or am I missing the boat somehow?

**Comment by rijkvanzanten** [comment]
Isn't that what hooks are, or am I misunderstanding the question? 🤔🤔

https://docs.directus.io/extensions/hooks/

**Reply by erondpowell** [reply]
I didn't know this existed yet 👏  Looks p much like what I meant.
I haven't read this thoroughly yet, but it appears this is only for custom extensions?

The only difference I imagined is making it easier by adding it to the GUI. You go to `data-flows > operations` then one of the operations is `custom function`. When you select it, the operation options menu has at least two form fields:
- One to pass in some parameter
- Another to set a function

These parameters and functions could be created right in the options menu, or there could be a separate Module/Interface.
Does that make sense? Or maybe it makes sense but it's not doable for some reason?

**Reply by rijkvanzanten** [reply]
At that point, you're just coding a new operation from the GUI. At that point, you're probably better of using an actual IDE / text editor 👍🏻 I don't really see any benefit to coding a custom hook from the app vs coding it outside of directus and putting the file in your extensions folder

**Reply by erondpowell** [reply]
Yes, coding a new operation within the GUI was my point.
I probably need to mess with this and just build one to see just how it works.

In my mind, the advantages would be:
1) to have it baked into the Data Flows workflow directly
2) be easily accessible via Directus Cloud

Perhaps these perceived advantages are actually a mirage? Seems like it by your last answer 😸

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
